File: adhan/src/test/java/com/batoulapps/adhan/TimingTest.java
package com.batoulapps.adhan;

import com.batoulapps.adhan.data.TimingFile;
import com.batoulapps.adhan.data.TimingInfo;
import com.batoulapps.adhan.data.TimingParameters;
import com.batoulapps.adhan.data.DateComponents;
import com.batoulapps.adhan.internal.TestUtils;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.io.FileFilter;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

import okio.Okio;

import static com.google.common.truth.Truth.assertThat;

public class TimingTest {

  private static final String PATH = "../Shared/Times/";

  private JsonAdapter<TimingFile> jsonAdapter;

  @Before
  public void setup() {
    Moshi moshi = new Moshi.Builder().build();
    jsonAdapter = moshi.adapter(TimingFile.class);
  }

  @Test
  public void testTimes() throws Exception {
    File timingDirectory = new File(PATH);
    File[] files = timingDirectory.listFiles(new FileFilter() {
      public boolean accept(File pathname) {
        return pathname.getName().endsWith(".json");
      }
    });

    for (File timingFile : files) {
      testTimingFile(timingFile);
    }
  }

  private void testTimingFile(File jsonFile) throws Exception {
    System.out.println("testing timings for " + jsonFile.getName());
    TimingFile timingFile = jsonAdapter.fromJson(Okio.buffer(Okio.source(jsonFile)));
    assertThat(timingFile).isNotNull();

    Coordinates coordinates = new Coordinates(
        timingFile.params.latitude, timingFile.params.longitude);
    CalculationParameters parameters = parseParameters(timingFile.params);

    for (TimingInfo info : timingFile.times) {
      DateComponents dateComponents = TestUtils.getDateComponents(info.date);
      PrayerTimes prayerTimes = new PrayerTimes(coordinates, dateComponents, parameters);
      long fajrDifference = getDifferenceInMinutes(prayerTimes.fajr, info.fajr, timingFile.params.timezone);
      assertThat(fajrDifference).isAtMost(timingFile.variance);
      long sunriseDifference = getDifferenceInMinutes(prayerTimes.sunrise, info.sunrise, timingFile.params.timezone);
      assertThat(sunriseDifference).isAtMost(timingFile.variance);
      long dhuhrDifference = getDifferenceInMinutes(prayerTimes.dhuhr, info.dhuhr, timingFile.params.timezone);
      assertThat(dhuhrDifference).isAtMost(timingFile.variance);
      long asrDifference = getDifferenceInMinutes(prayerTimes.asr, info.asr, timingFile.params.timezone);
      assertThat(asrDifference).isAtMost(timingFile.variance);
      long maghribDifference = getDifferenceInMinutes(prayerTimes.maghrib, info.maghrib, timingFile.params.timezone);
      assertThat(maghribDifference).isAtMost(timingFile.variance);
      long ishaDifference = getDifferenceInMinutes(prayerTimes.isha, info.isha, timingFile.params.timezone);
      assertThat(ishaDifference).isAtMost(timingFile.variance);
    }
  }

  private long getDifferenceInMinutes(Date prayerTime, String jsonTime, String timezone) {
    SimpleDateFormat formatter = new SimpleDateFormat("h:mm a");
    formatter.setTimeZone(TimeZone.getTimeZone(timezone));

    Date referenceTime;
    try {
      final Calendar parsedCalendar = Calendar.getInstance(TimeZone.getTimeZone(timezone));
      parsedCalendar.setTime(formatter.parse(jsonTime));

      final Calendar referenceCalendar = Calendar.getInstance(TimeZone.getTimeZone(timezone));
      referenceCalendar.setTime(prayerTime);
      referenceCalendar.set(Calendar.HOUR, parsedCalendar.get(Calendar.HOUR));
      referenceCalendar.set(Calendar.MINUTE, parsedCalendar.get(Calendar.MINUTE));
      referenceCalendar.set(Calendar.AM_PM, parsedCalendar.get(Calendar.AM_PM));
      referenceTime = referenceCalendar.getTime();
    } catch (Exception e) {
      referenceTime = new Date();
    }

    return Math.abs((prayerTime.getTime() - referenceTime.getTime()) / (60 * 1000));
  }

  private CalculationParameters parseParameters(TimingParameters timingParameters) {
    final CalculationMethod method;
    switch (timingParameters.method) {
      case "MuslimWorldLeague": {
        method = CalculationMethod.MUSLIM_WORLD_LEAGUE;
        break;
      }
      case "Egyptian": {
        method = CalculationMethod.EGYPTIAN;
        break;
      }
      case "Karachi": {
        method = CalculationMethod.KARACHI;
        break;
      }
      case "UmmAlQura": {
        method = CalculationMethod.UMM_AL_QURA;
        break;
      }
      case "Dubai": {
        method = CalculationMethod.DUBAI;
        break;
      }
      case "MoonsightingCommittee": {
        method = CalculationMethod.MOON_SIGHTING_COMMITTEE;
        break;
      }
      case "NorthAmerica": {
        method = CalculationMethod.NORTH_AMERICA;
        break;
      }
      case "Kuwait": {
        method = CalculationMethod.KUWAIT;
        break;
      }
      case "Qatar": {
        method = CalculationMethod.QATAR;
        break;
      }
      case "Singapore": {
        method = CalculationMethod.SINGAPORE;
        break;
      }
      default: {
        method = CalculationMethod.OTHER;
      }
    }

    CalculationParameters parameters = method.getParameters();
    if ("Shafi".equals(timingParameters.madhab)) {
      parameters.madhab = Madhab.SHAFI;
    } else if ("Hanafi".equals(timingParameters.madhab)) {
      parameters.madhab = Madhab.HANAFI;
    }

    if ("SeventhOfTheNight".equals(timingParameters.highLatitudeRule)) {
      parameters.highLatitudeRule = HighLatitudeRule.SEVENTH_OF_THE_NIGHT;
    } else if ("TwilightAngle".equals(timingParameters.highLatitudeRule)) {
      parameters.highLatitudeRule = HighLatitudeRule.TWILIGHT_ANGLE;
    } else {
      parameters.highLatitudeRule = HighLatitudeRule.MIDDLE_OF_THE_NIGHT;
    }

    return parameters;
  }

}

File: adhan/src/commonTest/kotlin/com/batoulapps/adhan/TimingTest.kt
package com.batoulapps.adhan

import com.batoulapps.adhan.CalculationMethod.DUBAI
import com.batoulapps.adhan.CalculationMethod.EGYPTIAN
import com.batoulapps.adhan.CalculationMethod.KARACHI
import com.batoulapps.adhan.CalculationMethod.KUWAIT
import com.batoulapps.adhan.CalculationMethod.MOON_SIGHTING_COMMITTEE
import com.batoulapps.adhan.CalculationMethod.MUSLIM_WORLD_LEAGUE
import com.batoulapps.adhan.CalculationMethod.NORTH_AMERICA
import com.batoulapps.adhan.CalculationMethod.OTHER
import com.batoulapps.adhan.CalculationMethod.QATAR
import com.batoulapps.adhan.CalculationMethod.SINGAPORE
import com.batoulapps.adhan.CalculationMethod.UMM_AL_QURA
import com.batoulapps.adhan.HighLatitudeRule.MIDDLE_OF_THE_NIGHT
import com.batoulapps.adhan.HighLatitudeRule.SEVENTH_OF_THE_NIGHT
import com.batoulapps.adhan.HighLatitudeRule.TWILIGHT_ANGLE
import com.batoulapps.adhan.Madhab.HANAFI
import com.batoulapps.adhan.Madhab.SHAFI
import com.batoulapps.adhan.data.DateComponents
import com.batoulapps.adhan.data.TimingFile
import com.batoulapps.adhan.data.TimingParameters
import com.batoulapps.adhan.internal.TestUtils.getDateComponents
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toInstant
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import okio.ExperimentalFileSystem
import okio.FileSystem
import okio.Path.Companion.toPath
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertTrue

class TimingTest {

  @ExperimentalFileSystem
  @Test
  fun testTimes() {
    val json = Json { ignoreUnknownKeys = true }
    val fs = FileSystem.SYSTEM

    val jsonPath = "../Shared/Times/".toPath()
    if (!fs.exists(jsonPath)) {
      // skip this test on iOS for now - should consider doing this via expect
      // and actual to avoid risking this not running on jvm, macOS, and others
      return
    }

    val dir = fs.list(jsonPath)
    dir.forEach { path ->
      val byteString = fs.read(path) {
        readByteString()
      }
      val contents = byteString.utf8()
      val timingFile = json.decodeFromString<TimingFile>(contents)
      validateTimingFile(timingFile)
    }
  }

  private fun validateTimingFile(timingFile: TimingFile) {
    val coordinates = Coordinates(timingFile.params.latitude, timingFile.params.longitude)
    val parameters = parseParameters(timingFile.params)

    for ((date, fajr, sunrise, dhuhr, asr, maghrib, isha) in timingFile.times) {
      val dateComponents: DateComponents = getDateComponents(date)
      val prayerTimes = PrayerTimes(coordinates, dateComponents, parameters)
      val fajrDifference =
        getDifferenceInMinutes(prayerTimes.fajr, fajr, timingFile.params.timezone)
      assertTrue { fajrDifference <= timingFile.variance }
      val sunriseDifference =
        getDifferenceInMinutes(prayerTimes.sunrise, sunrise, timingFile.params.timezone)
      assertTrue { sunriseDifference <= timingFile.variance }
      val dhuhrDifference =
        getDifferenceInMinutes(prayerTimes.dhuhr, dhuhr, timingFile.params.timezone)
      assertTrue { dhuhrDifference <= timingFile.variance }
      val asrDifference = getDifferenceInMinutes(prayerTimes.asr, asr, timingFile.params.timezone)
      assertTrue { asrDifference <= timingFile.variance }
      val maghribDifference =
        getDifferenceInMinutes(prayerTimes.maghrib, maghrib, timingFile.params.timezone)
      assertTrue { maghribDifference <= timingFile.variance }
      val ishaDifference =
        getDifferenceInMinutes(prayerTimes.isha, isha, timingFile.params.timezone)
      assertTrue { ishaDifference <= timingFile.variance }
    }
  }

  private fun getDifferenceInMinutes(
    prayerTime: Instant,
    jsonTime: String,
    timezone: String
  ): Long {
    val (time, amPm) = jsonTime.split(" ")
    val (hours, minutes) = time.split(":").map { it.toInt() }
    val resolvedHour = when {
      amPm == "PM" && hours == 12 -> 12
      amPm == "PM" -> hours + 12
      amPm == "AM" && hours == 12 -> 0
      else -> hours
    }

    val targetTimeZone = TimeZone.of(timezone)
    val calculatedDateTime = prayerTime.toLocalDateTime(targetTimeZone)

    val referenceDateTime = LocalDateTime(
      // PrayerTimes is in UTC, so we need calculatedDateTime here instead
      calculatedDateTime.year,
      calculatedDateTime.monthNumber,
      calculatedDateTime.dayOfMonth,
      resolvedHour,
      minutes
    )


    val referenceInstant = referenceDateTime.toInstant(targetTimeZone)
    val calculatedInstant = calculatedDateTime.toInstant(targetTimeZone)

    return abs(calculatedInstant.toEpochMilliseconds() - referenceInstant.toEpochMilliseconds()) / (60 * 1000)
  }

  private fun parseParameters(timingParameters: TimingParameters): CalculationParameters {
    val method: CalculationMethod = when (timingParameters.method) {
      "MuslimWorldLeague" -> MUSLIM_WORLD_LEAGUE
      "Egyptian" -> EGYPTIAN
      "Karachi" -> KARACHI
      "UmmAlQura" -> UMM_AL_QURA
      "Dubai" -> DUBAI
      "MoonsightingCommittee" -> MOON_SIGHTING_COMMITTEE
      "NorthAmerica" -> NORTH_AMERICA
      "Kuwait" -> KUWAIT
      "Qatar" -> QATAR
      "Singapore" -> SINGAPORE
      else -> OTHER
    }

    val parameters = method.parameters
    val madhab = if ("Shafi" == timingParameters.madhab) {
      SHAFI
    } else if ("Hanafi" == timingParameters.madhab) {
      HANAFI
    } else {
      parameters.madhab
    }

    val highLatitudeRule = if ("SeventhOfTheNight" == timingParameters.highLatitudeRule) {
      SEVENTH_OF_THE_NIGHT
    } else if ("TwilightAngle" == timingParameters.highLatitudeRule) {
      TWILIGHT_ANGLE
    } else {
      MIDDLE_OF_THE_NIGHT
    }
    return method.parameters.copy(madhab = madhab, highLatitudeRule = highLatitudeRule)
  }
}